Mocking Panache active record pattern not working

## Describe the bug
I've built a JAX-RS endpoint & a test using Panache & the active record pattern. I'm trying to mock Panache so I can test the resource class in isolation. It doesn't seem to be working. I've successfully been able to do this using the repository pattern and am trying to achieve the same thing with the active record pattern. I've attached a reproducing project, but here are the details.

I also thought that `void` methods on mocks do nothing by default (https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#doNothing--), so I tried eliminating the `PanacheMock.doNothing().when(Fruit.class).persist();` but that ended up with the same result.

Changing to use RESTEasy classic results in the same error.

I've also posted this on Zulip: https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/Need.20help.20mocking.20with.20Panache.20using.20active.20record.20pattern

JAX-RS class:

```java
@Path("/fruits")
public class FruitResource {
	@POST
	@Produces(MediaType.APPLICATION_JSON)
	@Consumes(MediaType.APPLICATION_JSON)
	@Blocking
	@Transactional
	public Fruit addFruit(@Valid Fruit fruit) {
		fruit.persist();
		return fruit;
	}
}
```

Test class

```java
@QuarkusTest
public class FruitResourceTests {
	@Test
	public void addFruit() {
		PanacheMock.mock(Fruit.class);
		PanacheMock.doNothing()
			.when(Fruit.class).persist();

		given()
			.when()
				.contentType(ContentType.JSON)
				.body("{\"id\":1,\"name\":\"Grapefruit\",\"description\":\"Summer fruit\"}")
				.post("/fruits")
			.then()
				.statusCode(200)
				.contentType(ContentType.JSON)
				.body(
					"id", is(1),
					"name", is("Grapefruit"),
					"description", is("Summer fruit")
				);

		PanacheMock.verify(Fruit.class).persist();
		PanacheMock.verifyNoMoreInteractions(Fruit.class);
	}
}
```

When running `./mvnw verify` on the attached reproducer I'm getting

```shell
2021-04-01 09:04:05,510 ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (executor-thread-1) Request failed: javax.persistence.PersistenceException: org.hibernate.PersistentObjectException: detached entity passed to persist: org.acme.domain.Fruit
        at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:154)
        at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:181)
        at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:188)
        at org.hibernate.internal.SessionImpl.firePersist(SessionImpl.java:726)
        at org.hibernate.internal.SessionImpl.persist(SessionImpl.java:706)
        at io.quarkus.hibernate.orm.runtime.session.TransactionScopedSession.persist(TransactionScopedSession.java:139)
        at io.quarkus.hibernate.orm.runtime.session.ForwardingSession.persist(ForwardingSession.java:53)
        at io.quarkus.hibernate.orm.panache.common.runtime.AbstractJpaOperations.persist(AbstractJpaOperations.java:99)
        at io.quarkus.hibernate.orm.panache.common.runtime.AbstractJpaOperations.persist(AbstractJpaOperations.java:94)
        at io.quarkus.hibernate.orm.panache.PanacheEntityBase.persist(PanacheEntityBase.java:56)
        at org.acme.rest.FruitResource.addFruit(FruitResource.java:23)
        at org.acme.rest.FruitResource_Subclass.addFruit$$superaccessor1(FruitResource_Subclass.zig:265)
        at org.acme.rest.FruitResource_Subclass$$function$$1.apply(FruitResource_Subclass$$function$$1.zig:33)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54)
        at io.quarkus.hibernate.validator.runtime.interceptor.AbstractMethodValidationInterceptor.validateMethodInvocation(AbstractMethodValidationInterceptor.java:68)
        at io.quarkus.hibernate.validator.runtime.jaxrs.QuarkusRestEndPointValidationInterceptor.validateMethodInvocation(QuarkusRestEndPointValidationInterceptor.java:19)
        at io.quarkus.hibernate.validator.runtime.jaxrs.QuarkusRestEndPointValidationInterceptor_Bean.intercept(QuarkusRestEndPointValidationInterceptor_Bean.zig:341)
        at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:50)
        at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:127)
        at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.invokeInOurTx(TransactionalInterceptorBase.java:100)
        at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.doIntercept(TransactionalInterceptorRequired.java:32)
        at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorBase.intercept(TransactionalInterceptorBase.java:53)
        at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired.intercept(TransactionalInterceptorRequired.java:26)
        at io.quarkus.narayana.jta.runtime.interceptor.TransactionalInterceptorRequired_Bean.intercept(TransactionalInterceptorRequired_Bean.zig:340)
        at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
        at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
        at org.acme.rest.FruitResource_Subclass.addFruit(FruitResource_Subclass.zig:222)
        at org.acme.rest.FruitResource$quarkusrestinvoker$addFruit_e4ceee061fa3e84c0954c49287079e5dcbfcc5b0.invoke(FruitResource$quarkusrestinvoker$addFruit_e4ceee061fa3e84c0954c49287079e5dcbfcc5b0.zig:39)
        at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
        at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:7)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:122)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2415)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at java.base/java.lang.Thread.run(Thread.java:834)
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Caused by: org.hibernate.PersistentObjectException: detached entity passed to persist: org.acme.domain.Fruit
        at org.hibernate.event.internal.DefaultPersistEventListener.onPersist(DefaultPersistEventListener.java:120)
        at org.hibernate.event.internal.DefaultPersistEventListener.onPersist(DefaultPersistEventListener.java:55)
        at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:93)
        at org.hibernate.internal.SessionImpl.firePersist(SessionImpl.java:720)
        ... 35 more

[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 13.704 s <<< FAILURE! - in org.acme.rest.FruitResourceTests

```

I also tried `PanacheMock.doNothing().when(Fruit.class).persist(Mockito.any(Fruit.class));` while changing the JAX-RS method to do `Fruit.persist(fruit);` instead of `fruit.persist();`, but that resulted in a much different error:

```shell
[ERROR] org.acme.rest.FruitResourceTests.addFruit  Time elapsed: 0.28 s  <<< ERROR!
java.lang.NullPointerException
        at org.acme.rest.FruitResourceTests.addFruit(FruitResourceTests.java:83)

[ERROR] org.acme.rest.FruitResourceTests.getAll  Time elapsed: 0.009 s  <<< ERROR!
org.mockito.exceptions.misusing.UnfinishedStubbingException:

Unfinished stubbing detected here:
-> at io.quarkus.panache.mock.PanacheMock.doNothing(PanacheMock.java:120)

E.g. thenReturn() may be missing.
Examples of correct stubbing:
    when(mock.isOk()).thenReturn(true);
    when(mock.isOk()).thenThrow(exception);
    doThrow(exception).when(mock).someVoidMethod();
Hints:
 1. missing thenReturn()
 2. you are trying to stub a final method, which is not supported
 3. you are stubbing the behaviour of another mock inside before 'thenReturn' instruction is completed

        at org.acme.rest.FruitResourceTests.getAll(FruitResourceTests.java:23)

2021-03-30 16:15:57,464 INFO  [io.quarkus] (main) Quarkus stopped in 0.631s
[INFO]
[INFO] Results:
[INFO]
[ERROR] Errors:
[ERROR]   FruitResourceTests.addFruit:83 » NullPointer
[ERROR]   FruitResourceTests.getAll:23 » UnfinishedStubbing
Unfinished stubbing detecte...
```

## To Reproduce
Reproducer project attached

Steps to reproduce the behavior:
1. Run `./mvnw verify`

### Configuration
```properties
# Add your application.properties here, if applicable.
quarkus.datasource.db-kind=postgresql
quarkus.datasource.username=fruits
quarkus.datasource.password=fruits
%prod.quarkus.datasource.jdbc.url=jdbc:postgresql://localhost:5432/fruits
quarkus.datasource.devservices.image-name=quay.io/edeandrea/postgres-13-fruits:latest

quarkus.hibernate-orm.database.generation=validate

# Set testcontainers info for testing
%test.quarkus.datasource.jdbc.url=jdbc:tc:postgresql:13:///test?TC_INITSCRIPT=db/schema.sql
%test.quarkus.datasource.jdbc.driver=org.testcontainers.jdbc.ContainerDatabaseDriver
```

## Environment (please complete the following information):

### Output of `uname -a` or `ver`
`Darwin edeandre-mac 20.3.0 Darwin Kernel Version 20.3.0: Thu Jan 21 00:07:06 PST 2021; root:xnu-7195.81.3~1/RELEASE_X86_64 x86_64`

### Output of `java -version`
```shell
openjdk version "11.0.2" 2019-01-15
OpenJDK Runtime Environment 18.9 (build 11.0.2+9)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.2+9, mixed mode)
```

### GraalVM version (if different from Java)

### Quarkus version or git rev
`1.13.0.Final` (I also tried with 1.12.x and same thing)

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)
```shell
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /Users/edeandre/.m2/wrapper/dists/apache-maven-3.6.3-bin/1iopthnavndlasol9gbrbg6bf2/apache-maven-3.6.3
Java version: 11.0.2, vendor: Oracle Corporation, runtime: /Users/edeandre/.sdkman/candidates/java/11.0.2-open
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "10.16", arch: "x86_64", family: "mac"
```

[code-with-quarkus.zip](https://github.com/quarkusio/quarkus/files/6243706/code-with-quarkus.zip)

This is what I meant - When I tried `PanacheMock.doNothing().when(Fruit.class).persist(Mockito.any(Fruit.class));` I got

```shell
[ERROR] org.acme.rest.FruitResourceTests.addFruit  Time elapsed: 0.28 s  <<< ERROR!
java.lang.NullPointerException
        at org.acme.rest.FruitResourceTests.addFruit(FruitResourceTests.java:83)

[ERROR] org.acme.rest.FruitResourceTests.getAll  Time elapsed: 0.009 s  <<< ERROR!
org.mockito.exceptions.misusing.UnfinishedStubbingException:

Unfinished stubbing detected here:
-> at io.quarkus.panache.mock.PanacheMock.doNothing(PanacheMock.java:120)

E.g. thenReturn() may be missing.
Examples of correct stubbing:
    when(mock.isOk()).thenReturn(true);
    when(mock.isOk()).thenThrow(exception);
    doThrow(exception).when(mock).someVoidMethod();
Hints:
 1. missing thenReturn()
 2. you are trying to stub a final method, which is not supported
 3. you are stubbing the behaviour of another mock inside before 'thenReturn' instruction is completed

        at org.acme.rest.FruitResourceTests.getAll(FruitResourceTests.java:23)

2021-03-30 16:15:57,464 INFO  [io.quarkus] (main) Quarkus stopped in 0.631s
[INFO]
[INFO] Results:
[INFO]
[ERROR] Errors:
[ERROR]   FruitResourceTests.addFruit:83 » NullPointer
[ERROR]   FruitResourceTests.getAll:23 » UnfinishedStubbing
Unfinished stubbing detecte...
```

--- a/extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/PanacheEntityBase.java
+++ b/extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/PanacheEntityBase.java
@@ -727,6 +727,7 @@ public abstract class PanacheEntityBase {
      * @see #persist(Stream)
      * @see #persist(Object,Object...)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static void persist(Iterable<?> entities) {
         JpaOperations.INSTANCE.persist(entities);
     }
@@ -739,6 +740,7 @@ public abstract class PanacheEntityBase {
      * @see #persist(Iterable)
      * @see #persist(Object,Object...)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static void persist(Stream<?> entities) {
         JpaOperations.INSTANCE.persist(entities);
     }
@@ -751,6 +753,7 @@ public abstract class PanacheEntityBase {
      * @see #persist(Stream)
      * @see #persist(Iterable)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static void persist(Object firstEntity, Object... entities) {
         JpaOperations.INSTANCE.persist(firstEntity, entities);
     }

--- a/extensions/panache/hibernate-reactive-panache/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/PanacheEntityBase.java
+++ b/extensions/panache/hibernate-reactive-panache/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/PanacheEntityBase.java
@@ -703,6 +703,7 @@ public abstract class PanacheEntityBase {
      * @see #persist(Stream)
      * @see #persist(Object,Object...)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static Uni<Void> persist(Iterable<?> entities) {
         return INSTANCE.persist(entities);
     }
@@ -716,6 +717,7 @@ public abstract class PanacheEntityBase {
      * @see #persist(Iterable)
      * @see #persist(Object,Object...)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static Uni<Void> persist(Stream<?> entities) {
         return INSTANCE.persist(entities);
     }
@@ -729,6 +731,7 @@ public abstract class PanacheEntityBase {
      * @see #persist(Stream)
      * @see #persist(Iterable)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static Uni<Void> persist(Object firstEntity, Object... entities) {
         return INSTANCE.persist(firstEntity, entities);
     }

--- a/extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/PanacheMongoEntityBase.java
+++ b/extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/PanacheMongoEntityBase.java
@@ -772,6 +772,7 @@ public abstract class PanacheMongoEntityBase {
      * @see #persist(Stream)
      * @see #persist(Object,Object...)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static void persist(Iterable<?> entities) {
         INSTANCE.persist(entities);
     }
@@ -784,6 +785,7 @@ public abstract class PanacheMongoEntityBase {
      * @see #persist(Iterable)
      * @see #persist(Object,Object...)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static void persist(Stream<?> entities) {
         INSTANCE.persist(entities);
     }
@@ -796,6 +798,7 @@ public abstract class PanacheMongoEntityBase {
      * @see #persist(Stream)
      * @see #persist(Iterable)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static void persist(Object firstEntity, Object... entities) {
         INSTANCE.persist(firstEntity, entities);
     }
@@ -808,6 +811,7 @@ public abstract class PanacheMongoEntityBase {
      * @see #update(Stream)
      * @see #update(Object,Object...)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static void update(Iterable<?> entities) {
         INSTANCE.update(entities);
     }
@@ -820,6 +824,7 @@ public abstract class PanacheMongoEntityBase {
      * @see #update(Iterable)
      * @see #update(Object,Object...)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static void update(Stream<?> entities) {
         INSTANCE.update(entities);
     }
@@ -832,6 +837,7 @@ public abstract class PanacheMongoEntityBase {
      * @see #update(Stream)
      * @see #update(Iterable)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static void update(Object firstEntity, Object... entities) {
         INSTANCE.update(firstEntity, entities);
     }
@@ -844,6 +850,7 @@ public abstract class PanacheMongoEntityBase {
      * @see #persistOrUpdate(Stream)
      * @see #persistOrUpdate(Object,Object...)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static void persistOrUpdate(Iterable<?> entities) {
         INSTANCE.persistOrUpdate(entities);
     }
@@ -856,6 +863,7 @@ public abstract class PanacheMongoEntityBase {
      * @see #persistOrUpdate(Iterable)
      * @see #persistOrUpdate(Object,Object...)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static void persistOrUpdate(Stream<?> entities) {
         INSTANCE.persistOrUpdate(entities);
     }
@@ -868,6 +876,7 @@ public abstract class PanacheMongoEntityBase {
      * @see #persistOrUpdate(Stream)
      * @see #persistOrUpdate(Iterable)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static void persistOrUpdate(Object firstEntity, Object... entities) {
         INSTANCE.persistOrUpdate(firstEntity, entities);
     }

--- a/extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/reactive/ReactivePanacheMongoEntityBase.java
+++ b/extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/reactive/ReactivePanacheMongoEntityBase.java
@@ -785,6 +785,7 @@ public abstract class ReactivePanacheMongoEntityBase {
      * @see #persist(Stream)
      * @see #persist(Object,Object...)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static Uni<Void> persist(Iterable<?> entities) {
         return INSTANCE.persist(entities);
     }
@@ -797,6 +798,7 @@ public abstract class ReactivePanacheMongoEntityBase {
      * @see #persist(Iterable)
      * @see #persist(Object,Object...)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static Uni<Void> persist(Stream<?> entities) {
         return INSTANCE.persist(entities);
     }
@@ -809,6 +811,7 @@ public abstract class ReactivePanacheMongoEntityBase {
      * @see #persist(Stream)
      * @see #persist(Iterable)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static Uni<Void> persist(Object firstEntity, Object... entities) {
         return INSTANCE.persist(firstEntity, entities);
     }
@@ -821,6 +824,7 @@ public abstract class ReactivePanacheMongoEntityBase {
      * @see #update(Stream)
      * @see #update(Object,Object...)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static Uni<Void> update(Iterable<?> entities) {
         return INSTANCE.update(entities);
     }
@@ -833,6 +837,7 @@ public abstract class ReactivePanacheMongoEntityBase {
      * @see #update(Iterable)
      * @see #update(Object,Object...)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static Uni<Void> update(Stream<?> entities) {
         return INSTANCE.update(entities);
     }
@@ -845,6 +850,7 @@ public abstract class ReactivePanacheMongoEntityBase {
      * @see #update(Stream)
      * @see #update(Iterable)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static Uni<Void> update(Object firstEntity, Object... entities) {
         return INSTANCE.update(firstEntity, entities);
     }
@@ -857,6 +863,7 @@ public abstract class ReactivePanacheMongoEntityBase {
      * @see #persistOrUpdate(Stream)
      * @see #persistOrUpdate(Object,Object...)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static Uni<Void> persistOrUpdate(Iterable<?> entities) {
         return INSTANCE.persistOrUpdate(entities);
     }
@@ -869,6 +876,7 @@ public abstract class ReactivePanacheMongoEntityBase {
      * @see #persistOrUpdate(Iterable)
      * @see #persistOrUpdate(Object,Object...)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static Uni<Void> persistOrUpdate(Stream<?> entities) {
         return INSTANCE.persistOrUpdate(entities);
     }
@@ -881,6 +889,7 @@ public abstract class ReactivePanacheMongoEntityBase {
      * @see #persistOrUpdate(Stream)
      * @see #persistOrUpdate(Iterable)
      */
+    @GenerateBridge(callSuperMethod = true)
     public static Uni<Void> persistOrUpdate(Object firstEntity, Object... entities) {
         return INSTANCE.persistOrUpdate(firstEntity, entities);
     }

--- a/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/visitors/PanacheEntityClassOperationGenerationVisitor.java
+++ b/extensions/panache/panache-common/deployment/src/main/java/io/quarkus/panache/common/deployment/visitors/PanacheEntityClassOperationGenerationVisitor.java
@@ -113,7 +113,7 @@ public class PanacheEntityClassOperationGenerationVisitor extends ClassVisitor {
             if (!userMethods.contains(method.name() + "/" + descriptor)) {
                 AnnotationInstance bridge = method.annotation(PanacheConstants.DOTNAME_GENERATE_BRIDGE);
                 if (bridge != null) {
-                    generateMethod(method, bridge.value("targetReturnTypeErased"));
+                    generateMethod(method, bridge.value("targetReturnTypeErased"), bridge.value("callSuperMethod"));
                 }
             }
         }
@@ -144,7 +144,7 @@ public class PanacheEntityClassOperationGenerationVisitor extends ClassVisitor {
         }
     }
 
-    protected void generateMethod(MethodInfo method, AnnotationValue targetReturnTypeErased) {
+    protected void generateMethod(MethodInfo method, AnnotationValue targetReturnTypeErased, AnnotationValue callSuperMethod) {
         List<org.jboss.jandex.Type> parameters = method.parameters();
 
         MethodVisitor mv = super.visitMethod(Opcodes.ACC_PUBLIC | Opcodes.ACC_STATIC | Opcodes.ACC_SYNTHETIC,
@@ -157,9 +157,17 @@ public class PanacheEntityClassOperationGenerationVisitor extends ClassVisitor {
         for (PanacheMethodCustomizer customizer : methodCustomizers) {
             customizer.customize(thisClass, method, mv);
         }
-        loadOperations(mv);
-        loadArguments(mv, parameters);
-        invokeOperations(mv, method);
+        if (callSuperMethod != null && callSuperMethod.asBoolean()) {
+            // delegate to super method
+            for (int i = 0; i < parameters.size(); i++) {
+                mv.visitIntInsn(Opcodes.ALOAD, i);
+            }
+            invokeOperations(mv, method, true);
+        } else {
+            loadOperations(mv);
+            loadArguments(mv, parameters);
+            invokeOperations(mv, method, false);
+        }
         mv.visitMaxs(0, 0);
         mv.visitEnd();
     }
@@ -177,11 +185,13 @@ public class PanacheEntityClassOperationGenerationVisitor extends ClassVisitor {
         }
     }
 
-    private void invokeOperations(MethodVisitor mv, MethodInfo method) {
+    private void invokeOperations(MethodVisitor mv, MethodInfo method, boolean callSuperMethod) {
         String operationDescriptor;
 
         StringJoiner joiner = new StringJoiner("", "(", ")");
-        joiner.add(CLASS.descriptor());
+        if (!callSuperMethod) {
+            joiner.add(CLASS.descriptor());
+        }
         descriptors(method, joiner);
 
         org.jboss.jandex.Type returnType = method.returnType();
@@ -191,9 +201,15 @@ public class PanacheEntityClassOperationGenerationVisitor extends ClassVisitor {
                 : returnType.name().toString();
         operationDescriptor = joiner + erasures.getOrDefault(key, descriptor);
 
-        mv.visitMethodInsn(INVOKEVIRTUAL, typeBundle.operations().internalName(), method.name(),
-                operationDescriptor, false);
-        if (returnType.kind() != org.jboss.jandex.Type.Kind.PRIMITIVE) {
+        if (callSuperMethod) {
+            mv.visitMethodInsn(Opcodes.INVOKESTATIC, typeBundle.entityBase().internalName(), method.name(),
+                    operationDescriptor, false);
+        } else {
+            mv.visitMethodInsn(INVOKEVIRTUAL, typeBundle.operations().internalName(), method.name(),
+                    operationDescriptor, false);
+        }
+        if (returnType.kind() != org.jboss.jandex.Type.Kind.PRIMITIVE
+                && returnType.kind() != org.jboss.jandex.Type.Kind.VOID) {
             String cast;
             if (returnType.kind() == org.jboss.jandex.Type.Kind.TYPE_VARIABLE) {
                 TypeVariable typeVariable = returnType.asTypeVariable();

--- a/extensions/panache/panache-common/runtime/src/main/java/io/quarkus/panache/common/impl/GenerateBridge.java
+++ b/extensions/panache/panache-common/runtime/src/main/java/io/quarkus/panache/common/impl/GenerateBridge.java
@@ -17,4 +17,9 @@ public @interface GenerateBridge {
      */
     boolean targetReturnTypeErased() default false;
 
+    /**
+     * Set to true to delegate to the super method instead of JpaOperations. This is useful to
+     * still inject interceptor calls and mock stubs.
+     */
+    boolean callSuperMethod() default false;
 }

--- a/integration-tests/hibernate-orm-panache/src/test/java/io/quarkus/it/panache/PanacheMockingTest.java
+++ b/integration-tests/hibernate-orm-panache/src/test/java/io/quarkus/it/panache/PanacheMockingTest.java
@@ -77,6 +77,9 @@ public class PanacheMockingTest {
         Assertions.assertSame(p, Person.findById(12l));
         Assertions.assertNull(Person.findById(42l));
 
+        Person.persist(p);
+        Assertions.assertNull(p.id);
+
         Mockito.when(Person.findById(12l)).thenThrow(new WebApplicationException());
         try {
             Person.findById(12l);
@@ -89,6 +92,7 @@ public class PanacheMockingTest {
 
         PanacheMock.verify(Person.class, Mockito.atLeast(5)).voidMethod();
         PanacheMock.verify(Person.class).findOrdered();
+        PanacheMock.verify(Person.class).persist(Mockito.<Object> any(), Mockito.<Object> any());
         PanacheMock.verify(Person.class, Mockito.atLeastOnce()).findById(Mockito.any());
         PanacheMock.verifyNoMoreInteractions(Person.class);
 
@@ -124,6 +128,9 @@ public class PanacheMockingTest {
         Assertions.assertSame(p, mockablePersonRepository.findById(12l));
         Assertions.assertNull(mockablePersonRepository.findById(42l));
 
+        mockablePersonRepository.persist(p);
+        Assertions.assertNull(p.id);
+
         Mockito.when(mockablePersonRepository.findById(12l)).thenThrow(new WebApplicationException());
         try {
             mockablePersonRepository.findById(12l);
@@ -136,6 +143,7 @@ public class PanacheMockingTest {
 
         Mockito.verify(mockablePersonRepository).findOrdered();
         Mockito.verify(mockablePersonRepository, Mockito.atLeastOnce()).findById(Mockito.any());
+        Mockito.verify(mockablePersonRepository).persist(Mockito.<Person> any());
         Mockito.verifyNoMoreInteractions(mockablePersonRepository);
     }
 

--- a/integration-tests/hibernate-reactive-panache/src/test/java/io/quarkus/it/panache/reactive/PanacheMockingTest.java
+++ b/integration-tests/hibernate-reactive-panache/src/test/java/io/quarkus/it/panache/reactive/PanacheMockingTest.java
@@ -43,6 +43,9 @@ public class PanacheMockingTest {
         Assertions.assertSame(p, Person.findById(12l).await().indefinitely());
         Assertions.assertNull(Person.findById(42l).await().indefinitely());
 
+        Person.persist(p).await().indefinitely();
+        Assertions.assertNull(p.id);
+
         Mockito.when(Person.findById(12l)).thenThrow(new WebApplicationException());
         try {
             Person.findById(12l);
@@ -54,6 +57,7 @@ public class PanacheMockingTest {
         Assertions.assertTrue(Person.findOrdered().await().indefinitely().isEmpty());
 
         PanacheMock.verify(Person.class).findOrdered();
+        PanacheMock.verify(Person.class).persist(Mockito.<Object> any(), Mockito.<Object> any());
         PanacheMock.verify(Person.class, Mockito.atLeastOnce()).findById(Mockito.any());
         PanacheMock.verifyNoMoreInteractions(Person.class);
 
@@ -89,6 +93,9 @@ public class PanacheMockingTest {
         Assertions.assertSame(p, mockablePersonRepository.findById(12l).await().indefinitely());
         Assertions.assertNull(mockablePersonRepository.findById(42l).await().indefinitely());
 
+        mockablePersonRepository.persist(p).await().indefinitely();
+        Assertions.assertNull(p.id);
+
         Mockito.when(mockablePersonRepository.findById(12l)).thenThrow(new WebApplicationException());
         try {
             mockablePersonRepository.findById(12l);
@@ -101,6 +108,7 @@ public class PanacheMockingTest {
 
         Mockito.verify(mockablePersonRepository).findOrdered();
         Mockito.verify(mockablePersonRepository, Mockito.atLeastOnce()).findById(Mockito.any());
+        Mockito.verify(mockablePersonRepository).persist(Mockito.<Person> any());
         Mockito.verifyNoMoreInteractions(mockablePersonRepository);
     }
 

--- a/integration-tests/mongodb-panache/src/main/java/io/quarkus/it/mongodb/panache/reactive/person/MockableReactivePersonRepository.java
+++ b/integration-tests/mongodb-panache/src/main/java/io/quarkus/it/mongodb/panache/reactive/person/MockableReactivePersonRepository.java
@@ -1,0 +1,16 @@
+package io.quarkus.it.mongodb.panache.reactive.person;
+
+import java.util.List;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import io.quarkus.mongodb.panache.reactive.ReactivePanacheMongoRepositoryBase;
+import io.quarkus.panache.common.Sort;
+import io.smallrye.mutiny.Uni;
+
+@ApplicationScoped
+public class MockableReactivePersonRepository implements ReactivePanacheMongoRepositoryBase<ReactivePersonEntity, Long> {
+    public Uni<List<ReactivePersonEntity>> findOrdered() {
+        return findAll(Sort.by("lastname", "firstname")).list();
+    }
+}

--- a/integration-tests/mongodb-panache/src/main/java/io/quarkus/it/mongodb/panache/reactive/person/ReactivePersonEntity.java
+++ b/integration-tests/mongodb-panache/src/main/java/io/quarkus/it/mongodb/panache/reactive/person/ReactivePersonEntity.java
@@ -1,12 +1,20 @@
 package io.quarkus.it.mongodb.panache.reactive.person;
 
+import java.util.List;
+
 import org.bson.codecs.pojo.annotations.BsonId;
 
 import io.quarkus.mongodb.panache.reactive.ReactivePanacheMongoEntityBase;
+import io.quarkus.panache.common.Sort;
+import io.smallrye.mutiny.Uni;
 
 public class ReactivePersonEntity extends ReactivePanacheMongoEntityBase {
     @BsonId
     public Long id;
     public String firstname;
     public String lastname;
+
+    public static Uni<List<ReactivePersonEntity>> findOrdered() {
+        return findAll(Sort.by("lastname", "firstname")).list();
+    }
 }

--- a/integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/MongodbPanacheMockingTest.java
+++ b/integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/MongodbPanacheMockingTest.java
@@ -48,6 +48,9 @@ public class MongodbPanacheMockingTest {
         Assertions.assertSame(p, PersonEntity.findById(12L));
         Assertions.assertNull(PersonEntity.findById(42L));
 
+        PersonEntity.persist(p);
+        Assertions.assertNull(p.id);
+
         Mockito.when(PersonEntity.findById(12L)).thenThrow(new WebApplicationException());
         try {
             PersonEntity.findById(12L);
@@ -59,6 +62,7 @@ public class MongodbPanacheMockingTest {
         Assertions.assertTrue(PersonEntity.findOrdered().isEmpty());
 
         PanacheMock.verify(PersonEntity.class).findOrdered();
+        PanacheMock.verify(PersonEntity.class).persist(Mockito.<Object> any(), Mockito.<Object> any());
         PanacheMock.verify(PersonEntity.class, Mockito.atLeastOnce()).findById(Mockito.any());
         PanacheMock.verifyNoMoreInteractions(PersonEntity.class);
     }
@@ -92,6 +96,9 @@ public class MongodbPanacheMockingTest {
         Assertions.assertSame(p, mockablePersonRepository.findById(12L));
         Assertions.assertNull(mockablePersonRepository.findById(42L));
 
+        mockablePersonRepository.persist(p);
+        Assertions.assertNull(p.id);
+
         Mockito.when(mockablePersonRepository.findById(12L)).thenThrow(new WebApplicationException());
         try {
             mockablePersonRepository.findById(12L);
@@ -104,6 +111,7 @@ public class MongodbPanacheMockingTest {
 
         Mockito.verify(mockablePersonRepository).findOrdered();
         Mockito.verify(mockablePersonRepository, Mockito.atLeastOnce()).findById(Mockito.any());
+        Mockito.verify(mockablePersonRepository).persist(Mockito.<PersonEntity> any());
         Mockito.verifyNoMoreInteractions(mockablePersonRepository);
     }
 

--- a/integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/reactive/ReactiveMongodbPanacheMockingTest.java
+++ b/integration-tests/mongodb-panache/src/test/java/io/quarkus/it/mongodb/panache/reactive/ReactiveMongodbPanacheMockingTest.java
@@ -1,0 +1,121 @@
+package io.quarkus.it.mongodb.panache.reactive;
+
+import java.time.Duration;
+import java.util.Collections;
+import java.util.List;
+
+import javax.ws.rs.WebApplicationException;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Order;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+
+import io.quarkus.it.mongodb.panache.MongoReplicaSetTestResource;
+import io.quarkus.it.mongodb.panache.reactive.person.MockableReactivePersonRepository;
+import io.quarkus.it.mongodb.panache.reactive.person.ReactivePersonEntity;
+import io.quarkus.panache.mock.PanacheMock;
+import io.quarkus.test.common.QuarkusTestResource;
+import io.quarkus.test.junit.QuarkusTest;
+import io.quarkus.test.junit.mockito.InjectMock;
+import io.smallrye.mutiny.Uni;
+
+@QuarkusTest
+@QuarkusTestResource(MongoReplicaSetTestResource.class)
+public class ReactiveMongodbPanacheMockingTest {
+
+    private static final Duration timeout = Duration.ofSeconds(2);
+
+    @Test
+    @Order(1)
+    public void testPanacheMocking() {
+        PanacheMock.mock(ReactivePersonEntity.class);
+
+        Assertions.assertEquals(0, ReactivePersonEntity.count().await().atMost(timeout));
+
+        Uni<Long> uni23 = Uni.createFrom().item(23L);
+        Mockito.when(ReactivePersonEntity.count()).thenReturn(uni23);
+        Assertions.assertEquals(uni23, ReactivePersonEntity.count());
+
+        Uni<Long> uni42 = Uni.createFrom().item(42L);
+        Mockito.when(ReactivePersonEntity.count()).thenReturn(uni42);
+        Assertions.assertEquals(uni42, ReactivePersonEntity.count());
+
+        Mockito.when(ReactivePersonEntity.count()).thenCallRealMethod();
+        Assertions.assertEquals(0, ReactivePersonEntity.count().await().atMost(timeout));
+
+        PanacheMock.verify(ReactivePersonEntity.class, Mockito.times(4)).count();
+
+        ReactivePersonEntity p = new ReactivePersonEntity();
+
+        Uni<ReactivePersonEntity> uniP = Uni.createFrom().item(p);
+        Mockito.when(ReactivePersonEntity.<ReactivePersonEntity> findById(12L)).thenReturn(uniP);
+        Assertions.assertSame(uniP, ReactivePersonEntity.findById(12L));
+        Assertions.assertNull(ReactivePersonEntity.findById(42L).await().atMost(timeout));
+
+        ReactivePersonEntity.persist(p);
+        Assertions.assertNull(p.id);
+
+        Mockito.when(ReactivePersonEntity.findById(12L)).thenThrow(new WebApplicationException());
+        try {
+            ReactivePersonEntity.findById(12L);
+            Assertions.fail();
+        } catch (WebApplicationException x) {
+        }
+
+        Uni<List<ReactivePersonEntity>> uniEmptyList = Uni.createFrom().item(Collections.emptyList());
+        Mockito.when(ReactivePersonEntity.findOrdered()).thenReturn(uniEmptyList);
+        Assertions.assertEquals(uniEmptyList, ReactivePersonEntity.findOrdered());
+
+        PanacheMock.verify(ReactivePersonEntity.class).findOrdered();
+        PanacheMock.verify(ReactivePersonEntity.class).persist(Mockito.<Object> any(), Mockito.<Object> any());
+        PanacheMock.verify(ReactivePersonEntity.class, Mockito.atLeastOnce()).findById(Mockito.any());
+        PanacheMock.verifyNoMoreInteractions(ReactivePersonEntity.class);
+    }
+
+    @InjectMock
+    MockableReactivePersonRepository mockablePersonRepository;
+
+    @Test
+    public void testPanacheRepositoryMocking() throws Throwable {
+        Assertions.assertEquals(0, mockablePersonRepository.count().await().atMost(timeout));
+
+        Uni<Long> uni23 = Uni.createFrom().item(23L);
+        Mockito.when(mockablePersonRepository.count()).thenReturn(uni23);
+        Assertions.assertEquals(uni23, mockablePersonRepository.count());
+
+        Uni<Long> uni42 = Uni.createFrom().item(42L);
+        Mockito.when(mockablePersonRepository.count()).thenReturn(uni42);
+        Assertions.assertEquals(uni42, mockablePersonRepository.count());
+
+        Mockito.when(mockablePersonRepository.count()).thenCallRealMethod();
+        Assertions.assertEquals(0, mockablePersonRepository.count().await().atMost(timeout));
+
+        Mockito.verify(mockablePersonRepository, Mockito.times(4)).count();
+
+        ReactivePersonEntity p = new ReactivePersonEntity();
+        Uni<ReactivePersonEntity> uniP = Uni.createFrom().item(p);
+        Mockito.when(mockablePersonRepository.findById(12L)).thenReturn(uniP);
+        Assertions.assertSame(uniP, mockablePersonRepository.findById(12L));
+        Assertions.assertNull(mockablePersonRepository.findById(42L).await().atMost(timeout));
+
+        mockablePersonRepository.persist(p);
+        Assertions.assertNull(p.id);
+
+        Mockito.when(mockablePersonRepository.findById(12L)).thenThrow(new WebApplicationException());
+        try {
+            mockablePersonRepository.findById(12L);
+            Assertions.fail();
+        } catch (WebApplicationException x) {
+        }
+
+        Uni<List<ReactivePersonEntity>> uniEmptyList = Uni.createFrom().item(Collections.emptyList());
+        Mockito.when(mockablePersonRepository.findOrdered()).thenReturn(uniEmptyList);
+        Assertions.assertEquals(uniEmptyList, mockablePersonRepository.findOrdered());
+
+        Mockito.verify(mockablePersonRepository).findOrdered();
+        Mockito.verify(mockablePersonRepository, Mockito.atLeastOnce()).findById(Mockito.any());
+        Mockito.verify(mockablePersonRepository).persist(Mockito.<ReactivePersonEntity> any());
+        Mockito.verifyNoMoreInteractions(mockablePersonRepository);
+    }
+}

